### DISCUSSION TITLE
Liquid based color matching demo

### DISCUSSION THREAD

**Discussion by sgbaird** [sparks-baird/self-driving-lab-demo#89] [Demo extensions and design]
Follow-up to https://github.com/sparks-baird/self-driving-lab-demo/discussions/77.

> ## Liquid-based color matching
> The demo I'm planning to design next (with similar instructions, bill of materials, etc.) is a color-matching demo using red/green/blue food coloring dissolved in water and housed in medical IV-style bags, pumped via several peristaltic pumps, with light matching via the existing AS7341 light sensor. I think I can design that one for under $200 and under 2 hours of setup time.

## Idea Summary

In principle, this is very similar to [Atinary SDLabs Demo](https://www.youtube.com/watch?v=SX26XRFx0U0) and a few other color-matching demos, but with a focus on spectrophotometry. Here is the visual summary:

<img src=https://github.com/sparks-baird/self-driving-lab-demo/raw/main/reports/figures/cls-lab-liquid-summary.png width=450>

### Drip bags
For the drip bags, polyolefin is probably the right chemical-resistant material, a combination of HDPE and PP.

- https://www.amazon.com/Drinks-Halloween-decorations-Pump-Efficient-Accessories-Halloween/dp/B0B5241PWK
  - <img src=https://user-images.githubusercontent.com/45469701/194430234-724e64dc-e461-4127-bc79-bb2cd98991a6.png width=200>
  - 350 mL
  - food grade PVC
- https://www.amazon.com/dp/B09723ZXD5
  - <img src=https://user-images.githubusercontent.com/45469701/194430343-da0c1fa9-1c28-4c75-87a9-b5a3b9140fef.png width=200>
  -  350 mL 
- https://www.amazon.com/HUHKOUAE-Drip-Bag-Plants-Irrigation/dp/B09TNJ75G6
  - <img src=https://user-images.githubusercontent.com/45469701/194433893-01208253-af3a-4eb2-84c8-82c7e08e1ff5.png width=300>
  - 3500 mL (large water bag, for plants)
  - 5 for $21.29
  - good for use as water and waste

### Cuvettes
- https://www.amazon.com/BrandTech-759071D-Polystyrene-2-5-4-5ml-Spectrophotometry/dp/B003UTUPPI
  - <img src=https://user-images.githubusercontent.com/45469701/194430595-40273e0d-bd51-4cd0-9689-8f92c77e1aea.png height=200>
  - no lid
  - poor transmission
  - $25.57 for pack of 100
- https://www.amazon.com/dp/B0829CD8SN/ref=emc_b_5_t
  - <img src=https://user-images.githubusercontent.com/45469701/194431824-ea4438f1-05c5-4cc1-a5a4-5b6b4edfe9e0.png width=200>
  - quartz
  - non-sealing PTFE lid
  - 3.5 mL
  - $39.99 for pack of 2

- https://www.amazon.com/dp/B07XZMMNT6/ref=emc_b_5_t
  - <img src=https://user-images.githubusercontent.com/45469701/194431786-bfaff22e-322c-44d9-85a3-58a0eb855e38.png width=200>
  - PTFE stopper top
  - 3.5 mL
  - $129.99 for pack of 2

- https://www.amazon.com/Quartz-Cuvette-Pack-Spectrophotometer-Transparent/dp/B086GVNK7X/
  - <img src=https://user-images.githubusercontent.com/45469701/194431438-a373560a-3829-4ae1-8546-9a61a6dfc21d.png width=200>
  - listed as quartz, but customer reported it being glass
  - sealable
  - 3.5 mL

- https://ecuvettes.com/product/qa25-silicone-cuvette-lid-with-center-hole/
  - Sealing lid with hole
  - $2

### Jugs

- https://smartbottleinc.com/product/5-3-gallon/Collapsible jug
  - https://smartbottleinc.com/product/5-3-gallon/

- https://www.uline.com/Product/Detail/S-15648/Jugs/Carboy-2-1-2-Gallon

### Food pouches
- https://www.amazon.com/dp/B08CC3NVKB
- https://www.amazon.com/iziusy-Reusable-Alcohol-Drinking-Concealable/dp/B07FBH592K (PE)

### Peristaltic pumps
- https://www.amazon.com/dp/B01IUVHB8E (12 V, but Maker Drive is 2.5V-9.5V)
- https://www.amazon.com/Kamoer-Peristaltic-Hydroponics-Nutrient-Analytical/dp/B07GWJ78FN/ (12 V)
- https://www.adafruit.com/product/3910
- https://www.amazon.com/Jebao-4-Channel-Saltwater-Aquarium-Doser3-4/dp/B08DJDH161
- Kamoer has another multi-channel option, but more expensive and does not have very good reviews it seems

## Motor Drivers

See https://github.com/CytronTechnologies/MAKER-PI-PICO/issues/13 for full suggestions.

- https://www.amazon.com/HiLetgo-H-bridge-Stepper-Controller-Arduino/dp/B00M0F243E
  - reviewer mentions that it pulls logic levels to 12V and could damage the microcontroller
  - reviewer mentions that there is no current protection on the outputs
  - low-cost, but various other issues
  - consistent with a lot of the cheap Amazon motor drivers
- [Maker Drive](https://www.cytron.io/p-maker-drive-simplifying-h-bridge-motor-driver-for-beginner)
  - 9.5V max, probably OK for driving 12V DC motors
- [DC Motor Driver Module for Raspberry Pi Pico](https://thepihut.com/products/dc-motor-driver-module-for-raspberry-pi-pico)
  -  6-12V input, benefit of having a breakout for all the pins
  - probably my first choice right now, might replace the use of Maker Pi Pico, or the Maker Pi Pico could be attached via (https://www.adafruit.com/product/4946 + 20-pin header)x2 
  - EDIT: great new product by Cytron, the "Robo Pico" https://www.cytron.io/p-robo-pico-simplifying-robotics-with-raspberry-pi-pico with 2 bidirectional DC channels, 4 servos, and 7 Grove ports, but only 3-6 V control.

### LEDs
- Maker Pi Pico RGB LED
- [Grove White LED](https://www.digikey.com/en/products/detail/seeed-technology-co.,-ltd/104030009/5487953)
  - able to replace with an arbitrary LED, which is appealing
  - the pin slots would be nice to use in a custom circuitboard, unless there's an existing product with a bunch of pin slots

### Drip Manifold

A drip manifold will combine multiple inlets and have them go to a single outlet; thus, only two ports are needed for the cuvette (see below). There needs to be at minimum 4 inlets, but preferably 5+ so that one can be used for gas (either open atmosphere or pulling from a sealed system). If using a sealed, atmospheric pressure system (i.e. the IV drip bag filled with air or gas), then an additional peristaltic pump is needed. If using a sealed, pressurized system (e.g. gas cylinder), then a solenoid valve and a constant-flow meter/controller (there are some cheap, manually adjustable ones) can be used, but then there needs to be an air outlet (e.g. bubbler) at some point downstream in the system.

Also called flow splitters.

some possible options (need to pick one and ensure appropriate connectors)

  - https://www.dripdepot.com/item/multiple-outlet-dripper-manifold-outlets-4-function-stacking
  - https://www.mcmaster.com/tube-fittings/shape~double-wye/connection-type~tube/for-use-with~beverage/ (white PP Double Wye reducer, note the tube hardness) or https://www.freshwatersystems.com/products/john-guest-three-way-divider-1-2-x-3-8 (2 of these can be stacked/connected to each other, more can be stacked if needed). I like this one because the shape should help prevent liquid from getting trapped which might lead to inaccurate data for the current experiment and might affect later experiments. The stackability means it's not limited to just 3 chemicals. If needing to use a Wye adapter instead of double Wye, then can use a 45 degree elbow to "correct" the bend if there's not one that is straight instead of Y-shaped.
  - https://www.amazon.com/Rake-shaped-discharge-Connector-Agricultural-Irrigation%EF%BC%888Pcs%EF%BC%89/dp/B09NL1V5X5/
- https://www.coleparmer.com/i/idex-low-pressure-manifold-assembly-natural-peek-5-port-0-062-bore-1-8-od-tubing-1-4-28-flat-bottom-1-ea/0202312
  - This is the type I've seen on other devices (e.g. chemputer)
- https://darwin-microfluidics.com/products/multiport-manifold-large-kit (similar to above)

A more expensive option would be [a multi-channel gradient valve](https://www.clippard.com/cms/wiki/brief-overview-gradient-valves) that I think would take the place of the manifold drip _and_ the peristaltic pumps.

Related: https://www.brechtel.com/product/flow-splitters/

## Theory

- [spectroscopy (wikipedia)](https://en.wikipedia.org/wiki/Spectroscopy)
- https://en.wikipedia.org/wiki/Ultraviolet%E2%80%93visible_spectroscopy

## Cuvette Ports

A flow cell cuvette would be ideal, e.g. [one from Starna](https://www.starnacells.com/d_cells_s/flow/xflow.html), but these are expensive. I could picture a formlabs 3D printed version using the resin I used for Devanathan electrochemical cells. See [this stack exchange post](https://chemistry.stackexchange.com/a/128576/89488) for context. The stopper cuvette seems viable; I could either modify the stopper to add two ports, or purchase a stopper with two ports already. It seems a bit unreasonable to expect having all 5 ports that would be necessary (3-colors, water rinse, waste removal). Mixing the solution can be done by using the waste pump cyclically in short spurts (simpler/easier than a stirring bar with a magnetic stirrer). Reminiscent of AM-ARES' use of a helical path insertion into a sponge to clean a needle tip (see one of the figures in SI of https://doi.org/10.1557/s43577-021-00051-1).

Probably some great things to be learned from C3PU by Cronin group:
<img src=https://user-images.githubusercontent.com/45469701/194436869-a3a8a8e7-53d3-44f3-8505-a07b46f5ccda.png width=450>

(see https://twitter.com/leecronin/status/1578042411061846021).

Maybe worth also asking Huat from the Guzik group (I'm thinking of the electrochemistry setup).

**Comment by sgbaird** [comment]
## Related Examples

- [Spectrophotometer Experiment Based on AS7341 Spectral Sensor](https://community.dfrobot.com/makelog-308416.html)
  - > In this experiment, I am going to use DFRobot AS7341 Spectral Sensor to make a simple spectrophotometer, and carry out an experiment to measure substance concentration in different solutions by collecting data, fitting curve, etc.
  - uses a red LED

- 3D printed fiber spectrometer kit
  - https://github.com/GaudiLabs/3DFiberSpectrometer/issues/3
  - https://gaudishop.ch/index.php/product/3d-printed-fiber-spectrometer-kit/
<img src=https://user-images.githubusercontent.com/45469701/195998885-160fb5c7-46df-4517-b468-7a7a73ce5e55.png width=300>
([image source](https://gaudishop.ch/wp-content/uploads/2021/03/Spectrometer_CaseOpen_Description.png))


## Theory
- https://en.wikipedia.org/wiki/Spectroscopy
- https://en.wikipedia.org/wiki/Ultraviolet%E2%80%93visible_spectroscopy
  - > UV/Vis can be applied to characterize the rate of a chemical reaction. Illustrative is the conversion of the yellow-orange and blue isomers of mercury dithizonate. (source)
  - > From the spectrum of burning gases, it is possible to determine a chemical composition of a fuel, temperature of gases, and air-fuel ratio.

The demo could be used to predict higher-fidelity UV-Vis spectra.

**Comment by sgbaird** [comment]
### Bill of Materials (WIP)

To be updated on main README once fleshed out.

- Everything from the CLSLab:Light demo [[DigiKey](https://www.digikey.com/short/045j7502)], ($60.80)
  - Pico W + headers (need to solder headers)
  - Maker Pi Pico base
  - Stemma-QT to Grove connector
  - AS7341 Light Sensor
  - USB-A to micro-USB-B cable
  - 5V USB-A wall adapter (might need one with a higher current rating, this one is rated at 5W)
  - 20 AWG sculpting wire
- Maker Drive x2 [[cytron](https://www.cytron.io/p-maker-drive-simplifying-h-bridge-motor-driver-for-beginner), 1A max current] ($4.50/ea., $9.00 total)
- Plant drip bag (large) x2 [[amazon](https://www.amazon.com/HUHKOUAE-Drip-Bag-Plants-Irrigation/dp/B09TNJ75G6)] ($21.29, semi-optional: can replace with extra 2 small bags from the party drink kit)
- Party drink drip bag (small) x3 [[amazon](https://www.amazon.com/Drinks-Halloween-decorations-Pump-Efficient-Accessories-Halloween/dp/B0B5241PWK)] ($29.99, includes pump and a 4-port water divider)
- Spectrophotometer cuvette [[amazon](https://www.amazon.com/Quartz-Cuvette-Pack-Spectrophotometer-Transparent/dp/B086GVNK7X/)] ($29.99)
- Peristaltic pump x4 [[amazon](https://www.amazon.com/dp/B01IUVHB8E)] ($11.98/ea., $47.92 total)
- silicone (or PTFE?) tubing (3 m) (what inner and outer diameter?) [[amazon](https://www.amazon.com/Teflon-tubing-Filament-Printer-Tech/dp/B073RDFTDV), 2 mm ID Teflon tubing] ($6.99), [[amazon](https://www.amazon.com/Silicone-Flexible-Rubber-Tubing-Transparent/dp/B08GPBC7C6/), 2mm ID silicone tubing, comes in other sizes] ($5.99) NOTE: 2 mm ID is for compatibility with [party drink setup](https://www.amazon.com/Drinks-Halloween-decorations-Pump-Efficient-Accessories-Halloween/dp/B0B5241PWK)
- LED light source (see https://github.com/sparks-baird/self-driving-lab-demo/discussions/52 for context), just use the existing Maker Pi Pico base RGB LED for now
- water pump (contained in the party drink kit)
- tube connectors (which kind?)
- double Wye connectors (just use the 4-port water divider from the party drink drip bag kit for now)
- silicone-to-glass glue or epoxy (to permanently fix a silicone lid to a cuvette) [[amazon](https://www.amazon.com/J-B-Weld-31310-All-Purpose-Silicone/dp/B00ID8EDKY/), JB-Weld] ($7.58)
- food dye (red, yellow, blue) [[amazon](https://www.amazon.com/Spice-Supreme-Assorted-Colors-Yellow/dp/B00GNHIBZU), $4.79]
- liquid-level chemical eTape [[milonetech](https://milonetech.com/products/chemical-etape-assembly)] ($59.99, optional)

**Reply by sgbaird** [reply]
BOM being updated at https://github.com/sparks-baird/self-driving-lab-demo/blob/main/reports/clslab-liquid-bom.xlsx

CSV (not necessarily always up-to-date) at https://github.com/sparks-baird/self-driving-lab-demo/blob/main/reports/clslab-liquid-bom.csv

**Comment by sgbaird** [comment]
## Chemically resistant replacements/modifications

- [Tygon chemical peristaltic tubing](https://www.processsystems.saint-gobain.com/products/tygon-chemical-peristaltic-pump-tubing) (and preferably PTFE when not in contact with the pump) instead of silicone
- Chemically resistant manifold, e.g.
  - https://www.coleparmer.com/i/idex-low-pressure-manifold-assembly-natural-peek-9-port-0-062-bore-1-8-od-tubing-1-4-28-flat-bottom-1-ea/0201368
  - https://www.mcmaster.com/ptfe-pipe-adapters/material~plastic/
- Replace plant IV drip bags with [medical-grade IV bags made of polyolefin](https://spiritmedical.com/products/iv-fluids-and-bags/easyflex-n-non-pvc/)
  - In the case of volatile liquids with high vapor pressure, a solution is to use the bottle it comes in directly, either upside down with a short needle poking through (and probably needing to be secured) or right-side up with a large needle reaching towards the bottom. The needle then needs to be connected directly to PTFE tubing or similar
- Replace the cuvette with a sealable flow-through cuvette (epoxy is probably not an option for sealing anymore)
  - https://www.wpiinc.com/cuv2062-1-flow-cell-small-volume-10mm-path-style-f
  - https://www.wpiinc.com/cuv2614-1-micro-volume-flow-quartz-cuvette-10mm-path-style-h
  - https://ecuvettes.com/product/qc3701-5-mm-flow-through-cuvette-2-clear-windows-650ul/
  - https://ecuvettes.com/product/qf88-10mm-1-3ml-air-tight-flow-through-cell-2-windows-powder-fused/
  - https://ecuvettes.com/product/qf95-10mm-pathlength-air-tight-flow-through-cell-27ml-2-clear-windows-3-m6-thread-connector-powder-fused/
  - https://ecuvettes.com/product/qf97-10mm-pathlength-3-5ml-flow-cell-2-clear-windows-uv-quartz-all-fused/
  - All fused is chemically resistant, high-transmission and powder-fused couldn't go quite as high in temperature, glued is not chemically resistant
  - https://www.fireflysci.com/flow-through-spectrophotometer-cells
  - https://www.amazon.com/Quartz-Flow-Cuvette-Flowing-Cuvettes/dp/B00GWDKN06
  - https://www.amazon.com/dp/B07SBZ9W2W

## Adding more PWM channels (i.e. more than 16)

Some options:

- https://www.adafruit.com/product/815
  - https://forum.arduino.cc/t/spi-and-pwm-from-the-same-pin-at-the-same-time/388389/6 (possible to do at separate times, not sure if this means reinstantiating classes or not)
- use https://www.waveshare.com/pico-motor-driver.htm with customizable I2C addresses
- use additional Pico W microcontrollers instead of trying to rely on a single one

**Comment by sgbaird** [comment]
## Task

Some thoughts after running an optimization campaign with the liquid color-matching demo:

- Higher food dye concentration is preferable
- Recording the background spectrum is probably a good idea

**Comment by iterateccvoelker** [comment]
This project is awesome again. It made me think about how cementitious materials are a lot like food dye - hear me out 🤣

It is a material that is usually characterized by spectroscopic methods such as XRF and XRF - It defies a complete characterization which is why experimental validation of the final product is always required. Here lies an optimization potential. The difficulty with cementitious materials is that amorphous and crystalline structures governs short-term and long-term behavior differently and are difficult to quantify distinctively - making validations an absolute must. Maybe you can introduce some sort of aging process in the color mixing 🤔.

In addition, the composition of alternative supplementary cementitious materials such as calcined clays or waste slags may be constantly changing - just as food dyes may not always have the same color when mixed.

They also present an optimization challenge that is highly dependent on (fully interchangeable) constituent information such as location, available volumes, eco-footprint, etc. Thus, almost endless optimization scenarios can be created.

I see many scenarios where food dye could act as some sort of surrogate material to develop clever optimization strategies for cement. However, unlike cement food dye has the great advantage of not being in powder form - one of the main drawbacks in the automated synthesis of cement.

I am thrilled to see how this project develops.

**Reply by sgbaird** [reply]
@BAMcvoelker thanks for the really interesting comments! This is good food for thought for me. I like the idea of incorporating aging in the color-mixing process. It's also good to know that the demo can capture some additional real-world application scenarios. Thanks for sharing!

**Comment by sgbaird** [comment]
Note: I made a working prototype, though it probably needs a 3D printed housing for the cuvette and could certainly be cleaned up regarding wires and connections.

![image](https://github.com/sparks-baird/self-driving-lab-demo/assets/45469701/a07854b4-5abd-4c12-a32a-f1f977f853f4)

https://github.com/sparks-baird/self-driving-lab-demo#clslabliquid

**Comment by sgbaird** [comment]
## Related endeavours
- PASCAL: https://github.com/fenning-research-group/PASCAL, https://www.youtube.com/watch?v=dKTjTsvgrgU, https://www.youtube.com/watch?v=l_-ezXn_QaU
- https://github.com/openUC2/UC2-MicroSyringePump
- Sidekick: https://www.sciencedirect.com/science/article/pii/S2468067222000645
- Bayesian optimization bartender https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7161969/
- OMIS: https://www.hardware-x.com/article/S2468-0672(18)30052-X/fulltext
- [Archerfish](https://www.linkedin.com/pulse/introducing-archerfish-direct-multi-material-printing-buonassisi)
- [Digital pipette](https://chemrxiv.org/engage/chemrxiv/article-details/6400e84e37e01856dc08de29)
